### PR TITLE
Feat - Resume training from a checkpoint

### PR DESCRIPTION
**Pull request by vincentlannelongue**
While retraining a model is already possible, all the training objects are reset (such as the learning rate). For longer trainings, it can be useful to resume a training with all the environment preserved, if the training is stopped for any reason.

The `resume_training` flag will distinguish 2 possibilities:
- if `True` - the training will continue training from the step when the model was saved
- if `False` - the training will start from step 0 but using the saved model

### CODE REVIEW

**Comment by vincentlannelongue** [comment]
NB: tried to resume the WandB run when resuming a training but this feature is unavailable when running in offline mode... https://github.com/wandb/wandb/issues/2423

**Comment by DonsetPG** [comment]
> NB: tried to resume the WandB run when resuming a training but this feature is unavailable when running in offline mode... [wandb/wandb#2423](https://github.com/wandb/wandb/issues/2423) For now, when resuming, a new WandB run is created, and the logged metrics start at training step 0 again, I'll see if there's something to do about this.

that's for the CICD right? or am i missing something?

**Comment by DonsetPG** [comment]
> Not sure how I can test the resume option in the CI/CD pipeline, if you guys have an idea?

do we have to? i'm fine with not testing it in the cicd

File: graphphysics/train.py [hunk 1]
```diff
@@ -36,7 +36,14 @@
...
 flags.DEFINE_integer("prefetch_factor", 2, "Number of batches to prefetch")
-flags.DEFINE_string("model_save_path", None, "Path to the checkpoint (.ckpt) file")
+flags.DEFINE_string(
+    "model_save_name", None, "Name to save the checkpoint during training"
+)
+flags.DEFINE_string(
+    "model_path", None, "Path to the checkpoint (.ckpt) to resume training from"
+)
+flags.DEFINE_bool("resume_training", False, "Whether to resume an unfinished training or not")
+
```
[the remark this answers is not in the document]
Comment (vincentlannelongue) [reply graphphysics/train.py:46]: The previous way of working when reloading model, typically if you want to finetune, and therefore restart the learning rate cycle from the beginning, this would be the way of doing so (starting back at epoch 0)

[the remark this answers is not in the document]
Comment (DonsetPG) [reply graphphysics/train.py:46]: fair ok

**Comment by vincentlannelongue** [comment]
> * for the cluster: can you try the `--append` argument when syncing? https://docs.wandb.ai/ref/cli/wandb-sync/ (unless i forgot how you send the logs from the cluster)

This might work if we specify the id of the first run when syncing the resumed run! 
If it does, I'll add at the end of the training log the ID of the first wandb run associated with the retrained model; this way we would just need to get this id and do  `wandb sync --append --id <run-id-from-first-run>`

File: graphphysics/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (theodore-michel) [reply graphphysics/train.py:46]: this is typically something i am doing already and want to keep as an option (e.g. train on a geometry, then load model to train on another type of geometry). For this it is preferable to start from scratch (epoch 0) with a user-defined warmup and learning rate which may differ from the training of the loaded model.

**Comment by theodore-michel** [comment]
> If it does, I'll add at the end of the training log the ID of the first wandb run associated with the retrained model; this way we would just need to get this id and do `wandb sync --append --id <run-id-from-first-run>`

I don't think you need to do that, seeing as at the end of the log there are 3 lines from wandb that give the ID, which you can directly use: 
`wandb: You can sync this run to the cloud by running:
wandb: wandb sync /path/to/run/wandb/offline-run-20250618_004037-cpmmc4cn
wandb: Find logs at: wandb/offline-run-20250618_004037-cpmmc4cn/logs`

**Comment by DonsetPG** [comment]
> > If it does, I'll add at the end of the training log the ID of the first wandb run associated with the retrained model; this way we would just need to get this id and do `wandb sync --append --id <run-id-from-first-run>`
> 
> I don't think you need to do that, seeing as at the end of the log there are 3 lines from wandb that give the ID, which you can directly use: `wandb: You can sync this run to the cloud by running:` `wandb: wandb sync /path/to/run/wandb/offline-run-20250618_004037-cpmmc4cn` `wandb: Find logs at: wandb/offline-run-20250618_004037-cpmmc4cn/logs`

not sure i understand what you mean?
(regarding the `--append`, people are suggesting that if you don't use this flag in offline mode, syncing with the same id overwrite previous logs)

**Comment by theodore-michel** [comment]
> not sure i understand what you mean?

My bad, i think i misinterpreted what @vincentlannelongue meant by adding the ID of the first run at the end of the log of the resumed training. However, i don't see how he can get the ID of the first run just with the .ckpt since the wandb run ID is not stored in it... for me this should just be something the user handles, best thing would be to print an info message at the end of the log saying that run should be synced by running `wandb sync --append --id <run-id-from-first-run>`.

**Comment by DonsetPG** [comment]
ah ok. Yeah that would work. Depending on what the cluster allow, we could also make an API call to fetch the latest run ID, but that's not necessarily the one we would want to resume. So yeah i think this should be handled by the user.

**Comment by vincentlannelongue** [comment]
> My bad, i think i misinterpreted what @vincentlannelongue meant by adding the ID of the first run at the end of the log of the resumed training. However, i don't see how he can get the ID of the first run just with the .ckpt since the wandb run ID is not stored in it... for me this should just be something the user handles, best thing would be to print an info message at the end of the log saying that run should be synced by running `wandb sync --append --id <run-id-from-first-run>`.

It's not a problem to store in the .ckpt a custom variable like `wandb_run_id`, which would allow when resuming training to have access to the ID of the WandB run ID corresponding to the first part of training. I could add something like "The first part of this training can be found at run <id>" somewhere in the log

**Comment by vincentlannelongue** [comment]
> My bad, i think i misinterpreted what @vincentlannelongue meant by adding the ID of the first run at the end of the log of the resumed training. However, i don't see how he can get the ID of the first run just with the .ckpt since the wandb run ID is not stored in it... for me this should just be something the user handles, best thing would be to print an info message at the end of the log saying that run should be synced by running `wandb sync --append --id <run-id-from-first-run>`.

It's doable to store the first WandB run ID in the .ckpt as a custom variable `wandb_run_id` that would be shown into the log when the `resume_training` flag is used

File: retrain.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+            --batch_size=1 \
+            --warmup=500 \
+            --num_workers=0 \
+            --prefetch_factor=0 \
+            --model_save_name=model \
+            --resume_training=True \
+            --model_path=checkpoints/model.ckpt \
+            --no_edge_feature \
+            --resume_training=True \
+            --model_path=checkpoints/model.ckpt
```
Comment (DonsetPG) [inline retrain.sh:14]: remove this

File: retrain.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+python -m graphphysics.train \
+            --training_parameters_path=mock_training.json \
+            --num_epochs=2 \
+            --init_lr=0.001 \
+            --batch_size=1 \
+            --warmup=500 \
+            --num_workers=0 \
+            --prefetch_factor=0 \
+            --model_save_name=model \
+            --resume_training=True \
```
Comment (DonsetPG) [inline retrain.sh:10]: should be `--resume_training`

File: graphphysics/train.py [hunk 4]
```diff
@@ -178,9 +190,21 @@ def main(argv):
...
+    if resume_training:
+        wandb_run = wandb.init(
+            project=wandb_project_name, id=lightning_module.wandb_run_id, resume="allow"
+        )
+    else:
+        wandb_run = wandb.init(project=wandb_project_name)
+
     wandb_logger = WandbLogger(experiment=wandb_run)
-    checkpoint_callback = ModelCheckpoint(dirpath="checkpoints/")
+    lightning_module.wandb_run_id = wandb_logger.experiment.id
```
Comment (DonsetPG) [inline graphphysics/train.py:201]: you tested this (with the callbacks) right? Can you also write a small unit test to test it?

**Review by DonsetPG — CHANGES_REQUESTED** [review]

File: retrain.sh (hunk 2 above)
Comment (vincentlannelongue) [reply retrain.sh:14]: Just saw your second comment, I guess you meant only the `=True`?

File: graphphysics/train.py (hunk 4 above)
Comment (DonsetPG) [reply graphphysics/train.py:201]: yes, that's it properly saved and loaded !

File: retrain.sh (hunk 2 above)
Comment (vincentlannelongue) [reply retrain.sh:14]: aaaaah my bad sorry 🤦
